### PR TITLE
Better keychain fails

### LINKED ISSUE #5199: ServiceRemoteXMLRPC causing crash with nil objects being inserted
Crash report: http://crashes.to/s/6f88b62a61c

Looks like we're trying to insert a nil object into an array and its causing a sporadic crash. Will probably do a 6.1.1 release to address this.

### PR DESCRIPTION
**Pull request by koke**
Now this would fix #5199 😉 

My goal with this PR is to not crash anywhere if a token or password is gone from the keychain. I think I've handled all the cases, but that doesn't mean it's a pretty UX. 

When a token/password is gone the app will not crash, but other ugly things will happen, like views "loading" forever, or controllers not being presented when you tap on a menu item.

The second part of this would address #687 and make sure that as soon as we know there's a bad auth state we modally prompt the user to re-authenticate.

For now we settle for not crashing, which at least gives users a chance to sign out without a reinstall.

I've cherry picked the commits from #5301, since `ServiceRemoteXMLRPC` now has a failable initializer, and I don't want that to get overwritten when 6.2 is merged into develop.

To test:

wp.com: Fresh install, login to WordPress.com, relaunch the app with `-WipeKeychainItem wordpress.com` (#5303), navigate around the app making sure there are no crashes.

Self hosted: Fresh install, add a self hosted, relaunch the app with `-WipeKeychainItem yoursite.com` (#5303), navigate around the app making sure there are no crashes.

Needs review: @astralbodies

### CODE REVIEW

File: WordPress/Classes/Services/AccountService.m [hunk 1]
```diff
@@ -256,6 +256,10 @@ - (void)updateUserDetailsForAccount:(WPAccount *)account success:(void (^)())suc
 
 - (id<AccountServiceRemote>)remoteForAccount:(WPAccount *)account
 {
+    if (account.restApi == nil) {
```
Comment (astralbodies) [inline WordPress/Classes/Services/AccountService.m]: Wow, I'm surprised this wasn't here already!

File: WordPress/Classes/Services/CommentService.m [hunk 2]
```diff
@@ -992,8 +992,10 @@ - (RemoteComment *)remoteCommentWithComment:(Comment *)comment
 {
     id<CommentServiceRemote>remote;
     // TODO: refactor API creation so it's not part of the model
-    if (blog.restApi) {
-        remote = [[CommentServiceRemoteREST alloc] initWithApi:blog.restApi siteID:blog.dotComID];
+    if ([blog supports:BlogFeatureWPComRESTAPI]) {
+        if (blog.restApi) {
+            remote = [[CommentServiceRemoteREST alloc] initWithApi:blog.restApi siteID:blog.dotComID];
```
Comment (astralbodies) [inline WordPress/Classes/Services/CommentService.m]: Couldn't the init method return nil if the API parameter was nil? The memory gets allocated but returning nil there frees it up right away, right?

File: WordPress/Classes/Services/AccountService.m (hunk 1 above)
Comment (koke) [inline WordPress/Classes/Services/AccountService.m]: I think before https://github.com/wordpress-mobile/WordPress-iOS/commit/977436bfdf8785ac47bcf001a0f46ffce30558fb `restApi` would never be `nil`, so it made no sense to check. 

I ❤️ nullability and optionals so much 😄

File: WordPress/Classes/Services/CommentService.m (hunk 2 above)
Comment (koke) [inline WordPress/Classes/Services/CommentService.m]: > Couldn't the init method return nil if the API parameter was nil?

I think it's a best design to ensure you're not passing a nil attribute, but that's actually a very good point, as ServiceRemoteXMLRPC uses the inverse approach. I should change that one instead.

I know this version is more verbose right now, but if you think about the future, would you declare the initializer in Swift as `init?(api: WordPressComApi?, siteID: Int?)`?

**Comment by astralbodies** [comment]
Did you look at https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/WordPress/Classes/ViewRelated/Stats/StatsViewController.m#L84 at all in your safety check? I'm fairly certain nothing crashes there but it might warrant updated logic for the `supports:` and Jetpack check.

File: WordPress/Classes/Services/CommentService.m (hunk 2 above)
Comment (astralbodies) [inline WordPress/Classes/Services/CommentService.m]: Totally agreed on the Swift `init?` vs Objective-C :smile:

Comment (koke) [inline WordPress/Classes/Services/CommentService.m]: Added safety checks for XML-RPC in 737fdd6

**Comment by koke** [comment]
> Did you look at StatsViewController

I did but not hard enough 😉 I saw this:

```
self.statsVC.oauth2Token = self.blog.restApi.authToken
```

and assumed the Stats VC would be fine with a `nil` token. I've tested it now and there's a failing assertion when initializing the `WPStatsService`, but it looks like Release builds won't crash

**Comment by astralbodies** [comment]
Cool - :shipit:!!

File: WordPress/Classes/Networking/ServiceRemoteXMLRPC.m [hunk 3]
```diff
@@ -13,12 +13,17 @@ @implementation ServiceRemoteXMLRPC
 
 - (id)initWithApi:(WPXMLRPCClient *)api username:(NSString *)username password:(NSString *)password
 {
+    NSParameterAssert(api != nil);
+    if (_username == nil || _password == nil) {
```
Comment (aerych) [inline WordPress/Classes/Networking/ServiceRemoteXMLRPC.m]: I think this was supposed to be `username` and `password`. Right now init will always return `nil`.

Comment (koke) [inline WordPress/Classes/Networking/ServiceRemoteXMLRPC.m]: Yep #5330 :scream:

Comment (aerych) [inline WordPress/Classes/Networking/ServiceRemoteXMLRPC.m]: I'm too slow and missed that it had already been patched ;)

**Pull request merged**